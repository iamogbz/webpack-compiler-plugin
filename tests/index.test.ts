import { Compiler } from "webpack";
import { WebpackCompilerPlugin } from "index";
import { logger } from "logger";

const processExitSpy = jest
    .spyOn(process, "exit")
    .mockImplementation(() => undefined as never);

const consoleLogSpy = jest.spyOn(logger, "info");
const consoleErrorSpy = jest.spyOn(logger, "error");

const mockCompiler = {
    hooks: {
        afterPlugins: { tap: jest.fn() },
        compilation: { tap: jest.fn() },
        done: { tap: jest.fn() },
    },
};

const newPlugin = ({
    listeners = {},
    stageMessages,
}: {
    listeners?: Partial<StageListeners>;
    stageMessages?: Partial<StageMessages>;
}) => {
    const plugin = new WebpackCompilerPlugin({
        name: "webpack-compiler-plugin",
        listeners,
        stageMessages,
    });
    plugin.apply((mockCompiler as unknown) as Compiler);
    return plugin;
};

afterEach(() => {
    jest.clearAllMocks();
    const signals = ["exit", "SIGINT", "uncaughtException"];
    signals.map(process.removeAllListeners.bind(process));
});
afterAll(jest.restoreAllMocks);

describe("default listeners", () => {
    it.each([
        ["interrupt", ["SIGINT"]],
        ["buildError", ["uncaughtException", new Error("Uncaught Error")]],
    ])("loads with default %s listener", (_, [event, arg]): void => {
        newPlugin({});
        process.emit(
            event as NodeJS.Signals,
            (arg as unknown) as NodeJS.Signals,
        );
        expect(consoleErrorSpy.mock.calls).toMatchSnapshot("console error");
        expect(consoleLogSpy.mock.calls).toMatchSnapshot("console log");
        expect(processExitSpy.mock.calls).toMatchSnapshot("process exit");
    });
});

describe("apply listeners", () => {
    it.each([
        ["buildEnd", ["exit"]],
        ["interrupt", ["SIGINT"]],
        ["buildError", ["uncaughtException", new Error("Uncaught Error")]],
    ])(
        "applies handler for stage %s",
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        (stage: Stage, [event, arg]: [NodeJS.Signals, any]) => {
            const mockHandler = jest.fn();
            newPlugin({ listeners: { [stage]: mockHandler } });
            process.emit(event, arg);
            expect(mockHandler).toHaveBeenCalledTimes(1);
        },
    );

    it.each([
        ["buildStart", mockCompiler.hooks.afterPlugins.tap],
        ["compileEnd", mockCompiler.hooks.done.tap],
        ["compileStart", mockCompiler.hooks.compilation.tap],
    ])("applies handler for stage %s", (stage, tapFn) => {
        const mockHandler = jest.fn();
        newPlugin({ listeners: { [stage as Stage]: mockHandler } });
        expect(tapFn).toHaveBeenCalledTimes(1);
        tapFn.mock.calls[0][1]();
        expect(mockHandler).toHaveBeenCalledTimes(1);
    });
});

describe("stage messages", () => {
    it.each([
        ["buildStart", mockCompiler.hooks.afterPlugins.tap],
        ["compileEnd", mockCompiler.hooks.done.tap],
        ["compileStart", mockCompiler.hooks.compilation.tap],
    ])("uses default state messages %s", (stage, tapFn) => {
        const mockHandler = jest.fn();
        newPlugin({ listeners: { [stage as Stage]: mockHandler } });
        tapFn.mock.calls[0][1]();
        expect(consoleLogSpy).toHaveBeenCalledTimes(1);
        expect(consoleLogSpy.mock.calls).toMatchSnapshot();
    });

    it.each([
        ["buildStart", mockCompiler.hooks.afterPlugins.tap],
        ["compileEnd", mockCompiler.hooks.done.tap],
        ["compileStart", mockCompiler.hooks.compilation.tap],
    ])("does not use default state messages %s", (stage, tapFn) => {
        const mockHandler = jest.fn();
        newPlugin({
            listeners: { [stage as Stage]: mockHandler },
            stageMessages: null,
        });
        tapFn.mock.calls[0][1]();
        expect(consoleLogSpy).toHaveBeenCalledTimes(0);
        expect(consoleLogSpy.mock.calls).toMatchSnapshot();
    });
});
