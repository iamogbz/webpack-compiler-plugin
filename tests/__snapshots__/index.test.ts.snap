// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default listeners loads with default buildError listener: console error 1`] = `
[
  [
    [Error: Uncaught Error],
  ],
]
`;

exports[`default listeners loads with default buildError listener: console log 1`] = `
[
  [
    "
[1mwebpack-compiler-plugin:[22m [31m🚒 build failed[39m",
  ],
]
`;

exports[`default listeners loads with default buildError listener: process exit 1`] = `
[
  [
    1,
  ],
]
`;

exports[`default listeners loads with default interrupt listener: console error 1`] = `[]`;

exports[`default listeners loads with default interrupt listener: console log 1`] = `
[
  [
    "
[1mwebpack-compiler-plugin:[22m [33m🚧 build interrupted[39m",
  ],
]
`;

exports[`default listeners loads with default interrupt listener: process exit 1`] = `
[
  [
    2,
  ],
]
`;

exports[`stage messages does not use default state messages buildStart 1`] = `[]`;

exports[`stage messages does not use default state messages compileEnd 1`] = `[]`;

exports[`stage messages does not use default state messages compileStart 1`] = `[]`;

exports[`stage messages uses default state messages buildStart 1`] = `
[
  [
    "
webpack-compiler-plugin: [34m🚀 build starting[39m",
  ],
]
`;

exports[`stage messages uses default state messages compileEnd 1`] = `
[
  [
    "
webpack-compiler-plugin: [32m⌛ code compiled[39m",
  ],
]
`;

exports[`stage messages uses default state messages compileStart 1`] = `
[
  [
    "
webpack-compiler-plugin: [34m⏳ code compiling[39m",
  ],
]
`;
